### PR TITLE
feat: enable explicit OSIDB lookup bypass for external CVE analysis

### PR DESCRIPTION
**Pull request by kdudka**
## What

Add per-request agent selection (`"agent": "public"`) to CVE analysis endpoints, allowing callers to analyse vulnerabilities not tracked in OSIDB — such as third-party advisories, embargoed CVEs, or dummy CVE IDs used during triage. The public agent removes OSIDB tools entirely and works only with CVE data supplied in the request body.

Also includes several fixes and improvements: `tools_used` is now populated from pydantic-ai's actual message history instead of relying on the LLM to self-report, a `_build_cve_input()` helper centralises prompt context construction with proper description precedence (`comment_zero` > `cve_description` > `description`), and the `SuggestCWE` prompt now disambiguates similar-sounding CWE weaknesses to improve classification accuracy.

## Why

PR #584 introduced mandatory OSIDB lookup, which breaks analysis of external vulnerabilities: dummy CVE IDs return 404, and user-provided descriptions get overwritten by OSIDB data. Rather than adding a runtime flag to suppress tool calls (while the LLM still sees them), switching to the `public_feature_agent` cleanly removes OSIDB tools from the LLM's view.

The `tools_used` fix addresses a reliability issue where the LLM would sometimes omit tools from its self-reported list or include `final_result` (pydantic-ai's synthetic structured-output tool). Extracting tool names from the actual `AgentRunResult` message history is both more accurate and simpler.

### Core features (`src/aegis_ai/`)

- **Per-request agent selection**: new `agent` field on CVE analysis requests selects `public` (no OSIDB tools) or `redhat` (with OSIDB tools); defaults to server-configured agent
- **`_build_cve_input()` helper**: extracted from inline logic in `SuggestImpact.exec()`; centralises prompt context construction with description precedence and forwards all OSIDB CVE fields (`title`, `statement`, `mitigation`, `comment_zero`, `comments`, `components`, `references`, `affects`, `cvss_scores`)
- **`CVEFeatureInput` model**: new Pydantic model carrying all CVE fields into the prompt, with a custom `__repr__` that omits `None` fields for cleaner logs
- **`_extract_tools_used()`**: new function that derives `tools_used` from pydantic-ai message history, filtering out `final_result` and deduplicating while preserving order
- **`SuggestCWE` prompt**: added disambiguation instruction for similar-sounding CWE weaknesses

### Web API (`src/aegis_ai_web/`)

- **`CVEAnalysisRequestBase`**: extracted shared fields (`cve_id`, `agent`) into a base model with `extra = "allow"` for OSIDB fields
- **`CVESingleAnalysisRequest`**: replaces raw `Request` in `POST /{feature}` handler, giving FastAPI proper request body validation and OpenAPI schema generation
- **OpenAPI specs**: updated `docs/openapi.json` and `docs/openapi.yml` to document the new `agent` field, OSIDB CVE fields, and mark request body as required

### Tests

- **`_build_cve_input` tests**: 8 tests covering all OSIDB field forwarding, description precedence chain, non-dict `static_context` handling, and `CVEFeatureInput.__repr__`
- **`_extract_tools_used` tests**: 5 tests covering tool extraction, `final_result` filtering, deduplication, and empty history
- **Web endpoint tests**: updated for Pydantic validation (422 instead of 400 for missing/invalid `cve_id`), added tests for `agent` field validation and public agent selection
- **`README.md`**: added REST API example for external CVE analysis with the public agent

### Evals

- **`ToolsUsedEvaluator`**: added `osv_dev_cve_tool` to accepted CVE data tools
- **Quality-review eval**: switched from static context to cached OSIDB tool for more realistic evaluation
- **GHSA cache**: added 40+ OSV.dev response fixtures for kernel and non-kernel CVEs
- **Known-to-fail annotations**: CVE-2024-44308 (`MitigationEvaluator`), CVE-2025-9222 (`StatementEvaluator`), CVE-2025-39795 (`StatementNoAffectsManifest`), CVE-2025-39810 (`CVSSKernelScopeAndPrivileges`), CVE-2026-23950 (`ComponentsOverlapEvaluator`)

## How to Test

```bash
# Start the dev server
make run-web

# Analyse an external CVE with the public agent
curl -s -X POST http://localhost:9000/api/v1/analysis/cve \
  -H 'Content-Type: application/json' \
  -d '{
    "cve_id": "CVE-2026-22100",
    "agent": "public",
    "title": "OCPP DataTransfer ReserveLogin command injection",
    "comment_zero": "The OCPP DataTransfer message ReserveLogin is vulnerable to command injection.",
    "impact": "CRITICAL",
    "components": ["ocpp-server"],
    "features": ["suggest-impact", "suggest-cwe"]
  }'

# Run the test suites
make check      # format + lint + type-check
make test       # unit tests
make test-web   # web API tests
```

## Related Tickets

Related: https://redhat.atlassian.net/browse/AEGIS-431

## Summary by Sourcery

Enable per-request agent selection for CVE analysis to support external/non-OSIDB vulnerabilities, improve prompt input handling, and make tool usage tracking and API validation more robust.

New Features:
- Add public vs Red Hat agent selection for CVE analysis requests, allowing OSIDB-free analysis of externally supplied CVE data.
- Introduce CVEFeatureInput and a shared _build_cve_input helper to funnel all CVE context fields into prompts with sensible description precedence.

Enhancements:
- Populate tools_used from pydantic-ai message history via _extract_tools_used instead of relying on LLM self-reporting.
- Refine the SuggestCWE prompt with explicit guidance to distinguish similar-sounding CWE weaknesses.
- Update single- and multi-CVE analysis endpoints to use Pydantic models, stricter CVEID validation, and OpenAPI docs that describe the request body, agent field, and OSIDB CVE fields.
- Adjust evals and quality-review tests to use cached CVE tooling and expand accepted CVE data tools, fixtures, and known-to-fail annotations.

Documentation:
- Document external CVE analysis using the public agent in README and update OpenAPI specs for CVE analysis requests.

Tests:
- Add unit tests for _build_cve_input and CVEFeatureInput repr, _extract_tools_used, agent resolution, and per-request agent selection in CVE endpoints, and update existing web endpoint tests for new validation behavior.

### CODE REVIEW

File: src/aegis_ai_web/src/main.py [hunk 1]
```diff
@@ -242,6 +242,14 @@ async def dispatch(self, request: Request, call_next):
     llm_agent = rh_feature_agent
 
 
+def _resolve_agent(agent_param: str | None):
+    if agent_param is None:
+        return llm_agent
+    if "public" in agent_param:
```
[the remark this answers is not in the document]
Comment (kdudka) [reply src/aegis_ai_web/src/main.py]: Fixed in 5957ec7d6b21d086affc3747d9f82b172ad7cf85.

File: src/aegis_ai_web/tests/test_endpoints.py [hunk 2]
```diff
@@ -611,3 +632,61 @@ def test_detail_flag(self):
...
+                    "cve_id": "CVE-2024-1234",
+                    "agent": "public",
+                    "title": "Test title",
+                    "cve_description": "Test description",
+                },
+            )
+        assert response.status_code == 200
+        # Verify the feature was constructed with the public agent
+        instance = mock_exec.call_args
+        assert instance is not None
```
[the remark this answers is not in the document]
Comment (kdudka) [reply src/aegis_ai_web/tests/test_endpoints.py]: Fixed in 232ac3c69222764df260243574205eb24adf6af0.

File: docs/openapi.yml [hunk 3]
```diff
@@ -136,6 +136,72 @@ paths:
...
+                agent:
+                  type: string
+                  enum:
+                  - public
+                  - redhat
+                  description: >-
+                    Agent to use for analysis. 'public' removes OSIDB tools so the
+                    LLM analyses user-provided text only. 'redhat' includes OSIDB
+                    tools (default when omitted; governed by AEGIS_WEB_FEATURE_AGENT).
+              additionalProperties: true
```
[the remark this answers is not in the document]
Comment (kdudka) [reply docs/openapi.yml:205]: The API is authenticated.  We currently do not set explicit bounds on the input fields.  The per-request LLM processing is limited by a timeout server-side.

File: docs/openapi.yml [hunk 4]
```diff
@@ -136,6 +136,72 @@ paths:
...
+          application/json:
+            schema:
+              type: object
+              required:
+              - cve_id
+              properties:
+                cve_id:
+                  type: string
+                  pattern: ^CVE-[0-9]{4}-[0-9]{4,7}$
+                  description: CVE identifier (e.g. CVE-2025-12345)
```
[the remark this answers is not in the document]
Comment (kdudka) [reply docs/openapi.yml:151]: This is gong to be reworked soon per https://redhat.atlassian.net/browse/AEGIS-245.

File: src/aegis_ai/osidb_bot/bot.py [hunk 5]
```diff
@@ -488,9 +493,11 @@ async def process_cve(self, cve: CVEID) -> None:
                 self.pending[flaw_updater.position()] = True  # mark as pending
             await flaw_updater.do()
 
-        except RuntimeError as e:
+        except BaseException as e:
             # something has failed
-            logger.warning(f"{cve}: {str(e)}")
+            handled_failure = isinstance(e, RuntimeError)
+            if handled_failure:
+                logger.warning(f"{cve}: {str(e)}")
```
[the remark this answers is not in the document]
Comment (kdudka) [reply src/aegis_ai/osidb_bot/bot.py:500]: This has already been waived in #748 (which has not yet been merged though).

File: docs/openapi.yml [hunk 6]
```diff
@@ -136,6 +136,72 @@ paths:
           type: boolean
           default: false
           title: Detail
+      requestBody:
+        content:
+          application/json:
+            schema:
+              type: object
+              required:
+              - cve_id
```
[the remark this answers is not in the document]
Comment (kdudka) [reply docs/openapi.yml:146]: Fixed in ed5ba264de138c3442a97467babe0a40e4e5f6da.

File: src/aegis_ai_web/tests/test_endpoints.py [hunk 7]
```diff
@@ -119,47 +127,56 @@ def test_yaml_openapi():
...
+    """
+    Test that empty 'cve_id' field in request body returns 422 (pattern mismatch)
+    instead of crashing with 500.
+    """
+    response = client.post(
+        "/api/v1/analysis/cve/suggest-impact",
+        json={"cve_id": ""},
+        headers={"Content-Type": "application/json"},
+    )
+    assert response.status_code == 422
```
[the remark this answers is not in the document]
Comment (kdudka) [reply src/aegis_ai_web/tests/test_endpoints.py:171]: Everything should be updated to document 422 in this case.  I can see no mention of 400 any more.

**Comment by kdudka** [comment]
@sourcery-ai review

File: src/aegis_ai/osidb_bot/bot.py [hunk 8]
```diff
@@ -488,9 +493,11 @@ async def process_cve(self, cve: CVEID) -> None:
...
 
             # schedule retry if enabled (but not for validation failures)
             if (
```
[the remark this answers is not in the document]
Comment (kdudka) [reply src/aegis_ai/osidb_bot/bot.py:503]: This is out of scope and already waived on #748.  The commit in question will be excluded from this PR before clearing the `draft` flag.

File: tests/test_features.py [hunk 9]
```diff
@@ -205,6 +205,66 @@ async def test_quality_review_with_test_model():
...
+class TestBuildCveInput:
+    """Tests for _build_cve_input() helper that populates prompt context."""
+
+    def test_no_static_context(self):
+        result = cve._build_cve_input("CVE-2025-1234")
+        assert result.cve_id == "CVE-2025-1234"
+        assert result.title is None
+        assert result.description is None
+        assert result.cwe_id is None
+        assert result.components is None
```
[the remark this answers is not in the document]
Comment (kdudka) [reply tests/test_features.py:219]: Fixed in 345b98aa500787521b906f34749a15627a0fb749.

File: docs/openapi.json [hunk 10]
```diff
@@ -217,6 +217,82 @@
...
+                    "type": "string",
+                    "enum": ["public", "redhat"],
+                    "description": "Agent to use for analysis. 'public' removes OSIDB tools so the LLM analyses user-provided text only. 'redhat' includes OSIDB tools (default when omitted; governed by AEGIS_WEB_FEATURE_AGENT)."
+                  }
+                },
+                "additionalProperties": true
+              }
+            }
+          }
+        },
```
[the remark this answers is not in the document]
Comment (kdudka) [reply docs/openapi.json:295]: This was a misleading commit message only.  Should be fixed now.

File: tests/test_features.py [hunk 11]
```diff
@@ -205,6 +205,66 @@ async def test_quality_review_with_test_model():
     assert len(quality_review.manual_context_needed) > 0
```
[the remark this answers is not in the document]
Comment (kdudka) [reply tests/test_features.py:207]: Fixed in bb593011c40688cafbc8dec54866c6721d7c548d.

File: docs/openapi.json [hunk 12]
```diff
@@ -598,6 +674,19 @@
...
+              {
+                "type": "string",
+                "enum": ["public", "redhat"]
+              },
+              {
+                "type": "null"
+              }
+            ],
+            "title": "Agent",
+            "description": "Agent to use for analysis. 'public' removes OSIDB tools so the LLM analyses user-provided text only. 'redhat' includes OSIDB tools (default when omitted; governed by AEGIS_WEB_FEATURE_AGENT)."
```
[the remark this answers is not in the document]
Comment (kdudka) [reply docs/openapi.json:689]: `osidb_tool` is not the only source of input data.  There are other tools available (also) to the `public` agent.

File: evals/ghsa_cache/CVE-2022-50235.json [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+                    "type": "GIT",
+                    "repo": "https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git",
+                    "events": [
+                        {
+                            "introduced": "5.15.12"
+                        },
+                        {
+                            "fixed": "5.15.75"
+                        }
+                    ]
```
[the remark this answers is not in the document]
Comment (kdudka) [reply evals/ghsa_cache/CVE-2022-50235.json:109]: This is just cached external data used by our evaluation suite.

File: evals/ghsa_cache/CVE-2026-26740.json [hunk 14]
```diff
@@ -0,0 +1 @@
+{}
```
[the remark this answers is not in the document]
Comment (kdudka) [reply evals/ghsa_cache/CVE-2026-26740.json:1]: This is how all https://api.osv.dev/v1/vulns/CVE-YYYY-NNNNNN URLs returning 404 are cached.

File: src/aegis_ai_web/src/main.py [hunk 15]
```diff
@@ -516,34 +534,31 @@ async def _run_feature(feature_name: str) -> Any:
...
     FeatureClass = cve_feature_registry[feature_name]
-    try:
-        validated_input = dict(cve_data)
-    except Exception as e:
-        msg = f"Invalid input for CVE feature '{feature_name}'"
-        log_exception_safely(e, msg)
-        raise HTTPException(status_code=422, detail=msg)
+
+    validated_input: dict[str, Any] = {"cve_id": cve_id}
+    if request_body.model_extra:
```
[the remark this answers is not in the document]
Comment (kdudka) [reply src/aegis_ai_web/src/main.py:550]: I would not over-complicate this for now.
It might need to be reworked because of https://redhat.atlassian.net/browse/AEGIS-245 soon anyway.

File: src/aegis_ai/features/__init__.py [hunk 16]
```diff
@@ -9,11 +9,29 @@
...
+def _extract_tools_used(result: AgentRunResult) -> list[str]:
+    """Extract unique tool names from the pydantic-ai message history.
+
+    Excludes pydantic-ai's synthetic ``final_result`` tool used internally
+    for structured output delivery.
+    """
+    return list(
+        dict.fromkeys(
+            tc.tool_name
+            for msg in result.all_messages()
```
[the remark this answers is not in the document]
Comment (kdudka) [reply src/aegis_ai/features/__init__.py:27]: Fixed in 44d26bd645012e938d9c4c6990f511960b2a256d.

File: tests/test_features.py [hunk 17]
```diff
@@ -205,6 +207,158 @@ async def test_quality_review_with_test_model():
...
+        assert result.description == "NVD description text."
+
+    def test_description_precedence_comment_zero_wins(self):
+        ctx = {
+            "comment_zero": "from comment_zero",
+            "cve_description": "from cve_description",
+            "description": "from description",
+        }
+        result = cve._build_cve_input("CVE-2025-1234", ctx)
+        assert result.description == "from comment_zero"
```
[the remark this answers is not in the document]
Comment (kdudka) [reply tests/test_features.py:268]: The test coverage is sufficient for now.  This might need to be reworked soon because of https://redhat.atlassian.net/browse/AEGIS-245 anyway.

File: tests/test_features.py [hunk 18]
```diff
@@ -205,6 +207,158 @@ async def test_quality_review_with_test_model():
...
+class TestExtractToolsUsed:
+    """Tests for _extract_tools_used() that derives tools_used from message history."""
+
+    def test_extracts_tool_names_in_order(self):
+        result = _fake_agent_run_result(["osidb_tool", "cwe_tool"])
+        assert _extract_tools_used(result) == ["osidb_tool", "cwe_tool"]
+
+    def test_filters_out_final_result(self):
+        result = _fake_agent_run_result(["osidb_tool", "final_result"])
+        assert _extract_tools_used(result) == ["osidb_tool"]
```
[the remark this answers is not in the document]
Comment (kdudka) [reply tests/test_features.py:347]: This is already sufficiently covered by the evaluation suite.

File: src/aegis_ai_web/tests/test_endpoints.py [hunk 19]
```diff
@@ -611,3 +621,103 @@ def test_detail_flag(self):
...
+class TestResolveAgent:
+    """Tests for per-request agent selection via _resolve_agent()."""
+
+    def test_none_returns_default(self):
+        assert _resolve_agent(None) is llm_agent
+
+    def test_public_returns_public_agent(self):
+        assert _resolve_agent("public") is public_feature_agent
+
+    def test_redhat_returns_rh_agent(self):
```
[the remark this answers is not in the document]
Comment (kdudka) [reply src/aegis_ai_web/tests/test_endpoints.py:635]: This is a corner case really, no need for test coverage IMO.

File: src/aegis_ai_web/tests/test_endpoints.py [hunk 20]
```diff
@@ -611,3 +621,103 @@ def test_detail_flag(self):
...
+class TestAgentSelectionEndpoints:
+    """Tests for per-request agent selection in POST CVE analysis endpoints."""
+
+    def test_post_single_feature_with_public_agent(self):
+        """POST /api/v1/analysis/cve/{feature} respects agent='public'."""
+        mock_result = _make_mock_result(_make_sac_output())
+        captured_agents = []
+
+        def spy_init(self, agent):
+            captured_agents.append(agent)
```
[the remark this answers is not in the document]
Comment (kdudka) [reply src/aegis_ai_web/tests/test_endpoints.py:659]: Sounds over complicated.  We would spend much more time on maintaining the tests than the implementation itself.

**Review by superbuggy — APPROVED** [review]
